File: backend/tests/test_types.rs
// Simple example to test the core data types and configuration system
use file_transfer_backend::{TransferConfig, Protocol, TransferMode, TransferProgress, TransferResult, TransferStatus};
use std::time::Duration;

fn main() {
    println!("Testing core data types and configuration system...");
    
    // Test Protocol enum with FromStr
    let tcp_protocol: Protocol = "tcp".parse().expect("Failed to parse TCP protocol");
    let udp_protocol: Protocol = "UDP".parse().expect("Failed to parse UDP protocol");
    println!("✓ Protocol parsing works: {} and {}", tcp_protocol, udp_protocol);
    
    // Test TransferMode enum with FromStr
    let transmitter_mode: TransferMode = "transmitter".parse().expect("Failed to parse transmitter mode");
    let receiver_mode: TransferMode = "receiver".parse().expect("Failed to parse receiver mode");
    println!("✓ TransferMode parsing works: {} and {}", transmitter_mode, receiver_mode);
    
    // Test TransferConfig creation and validation
    let config = TransferConfig::new(
        TransferMode::Transmitter,
        Protocol::Tcp,
        Some("192.168.1.1".to_string()),
        8080,
        Some("test.txt".to_string()),
        Some(4096),
        Some(Duration::from_secs(60)),
    ).expect("Failed to create valid config");
    
    println!("✓ TransferConfig creation and validation works");
    println!("  - Mode: {}", config.mode);
    println!("  - Protocol: {}", config.protocol);
    println!("  - Target IP: {:?}", config.target_ip);
    println!("  - Port: {}", config.port);
    println!("  - Chunk size: {}", config.chunk_size);
    println!("  - Timeout: {:?}", config.timeout);
    
    // Test default configuration
    let default_config = TransferConfig::default();
    println!("✓ Default configuration works");
    println!("  - Default mode: {}", default_config.mode);
    println!("  - Default protocol: {}", default_config.protocol);
    println!("  - Default port: {}", default_config.port);
    
    // Test TransferProgress
    let mut progress = TransferProgress::new("test-transfer-123".to_string());
    progress.update(0.75, 1024.0 * 1024.0, 30); // 75% complete, 1MB/s, 30s remaining
    
    println!("✓ TransferProgress works");
    println!("  - Transfer ID: {}", progress.transfer_id);
    println!("  - Progress: {:.1}%", progress.progress_percentage());
    println!("  - Speed: {}", progress.speed_human_readable());
    println!("  - ETA: {}", progress.eta_human_readable());
    println!("  - Status: {}", progress.status);
    
    // Test TransferStatus methods
    println!("✓ TransferStatus methods work");
    println!("  - Idle is terminal: {}", TransferStatus::Idle.is_terminal());
    println!("  - Completed is terminal: {}", TransferStatus::Completed.is_terminal());
    println!("  - Connecting is active: {}", TransferStatus::Connecting.is_active());
    println!("  - Error is active: {}", TransferStatus::Error.is_active());
    
    // Test TransferResult
    let result = TransferResult::success(
        "test-transfer-123".to_string(),
        1024 * 1024 * 10, // 10 MB
        Duration::from_secs(60), // 1 minute
        "abc123def456".to_string(),
    );
    
    println!("✓ TransferResult works");
    println!("  - Success: {}", result.success);
    println!("  - Bytes transferred: {}", result.bytes_human_readable());
    println!("  - Average speed: {}", result.speed_human_readable());
    println!("  - Duration: {:?}", result.duration);
    println!("  - Checksum: {}", result.checksum);
    
    // Test configuration validation errors
    let invalid_config = TransferConfig::new(
        TransferMode::Transmitter,
        Protocol::Tcp,
        None, // Missing target IP for transmitter
        8080,
        Some("test.txt".to_string()),
        Some(4096),
        Some(Duration::from_secs(60)),
    );
    
    match invalid_config {
        Ok(_) => println!("✗ Configuration validation failed - should have rejected missing target IP"),
        Err(e) => println!("✓ Configuration validation works: {}", e),
    }
    
    println!("\n🎉 All core data types and configuration system tests passed!");
}

File: frontend/src/lib/test-types.ts
// Simple test to verify TypeScript types work correctly
import { 
    TransferConfigValidator, 
    TransferUtils,
    defaultTransferConfig,
    type TransferConfig,
    type TransferProgress, 
    type TransferResult, 
    type TransferStatus,
    type Protocol,
    type TransferMode
} from './types';

// Test function to verify all types work
export function testTypes(): void {
    console.log('Testing TypeScript types and utilities...');
    
    // Test default configuration
    const config: TransferConfig = { ...defaultTransferConfig };
    config.targetIp = '192.168.1.1';
    config.port = 8080;
    config.protocol = 'tcp';
    config.mode = 'transmitter';
    
    console.log('✓ TransferConfig type works');
    console.log('  - Mode:', config.mode);
    console.log('  - Protocol:', config.protocol);
    console.log('  - Target IP:', config.targetIp);
    console.log('  - Port:', config.port);
    
    // Test configuration validation
    const validationErrors = TransferConfigValidator.validate(config);
    if (validationErrors.length === 0) {
        console.log('✓ Configuration validation works - no errors');
    } else {
        console.log('✗ Configuration validation failed:', validationErrors);
    }
    
    // Test invalid configuration
    const invalidConfig: TransferConfig = {
        ...defaultTransferConfig,
        mode: 'transmitter',
        targetIp: undefined, // Missing required field
        port: 70000, // Invalid port
    };
    
    const invalidErrors = TransferConfigValidator.validate(invalidConfig);
    if (invalidErrors.length > 0) {
        console.log('✓ Configuration validation catches errors:', invalidErrors);
    } else {
        console.log('✗ Configuration validation should have caught errors');
    }
    
    // Test TransferProgress type
    const progress: TransferProgress = {
        transferId: 'test-123',
        progress: 0.75,
        speed: 1024 * 1024, // 1 MB/s
        eta: 30,
        status: 'transferring',
    };
    
    console.log('✓ TransferProgress type works');
    console.log('  - Transfer ID:', progress.transferId);
    console.log('  - Progress:', TransferUtils.formatProgress(progress.progress));
    console.log('  - Speed:', TransferUtils.formatSpeed(progress.speed));
    console.log('  - ETA:', TransferUtils.formatDuration(progress.eta));
    console.log('  - Status:', progress.status);
    
    // Test TransferResult type
    const result: TransferResult = {
        success: true,
        transferId: 'test-123',
        bytesTransferred: 10 * 1024 * 1024, // 10 MB
        duration: 60, // 1 minute
        checksum: 'abc123def456',
    };
    
    console.log('✓ TransferResult type works');
    console.log('  - Success:', result.success);
    console.log('  - Bytes transferred:', TransferUtils.formatBytes(result.bytesTransferred));
    console.log('  - Duration:', TransferUtils.formatDuration(result.duration));
    console.log('  - Checksum:', result.checksum);
    
    // Test utility functions
    console.log('✓ TransferUtils work');
    console.log('  - Format bytes (1024):', TransferUtils.formatBytes(1024));
    console.log('  - Format bytes (1536KB):', TransferUtils.formatBytes(1536 * 1024));
    console.log('  - Format bytes (2.5MB):', TransferUtils.formatBytes(2.5 * 1024 * 1024));
    console.log('  - Format speed (1MB/s):', TransferUtils.formatSpeed(1024 * 1024));
    console.log('  - Format duration (90s):', TransferUtils.formatDuration(90));
    console.log('  - Format duration (3665s):', TransferUtils.formatDuration(3665));
    
    // Test status utility functions
    const statuses: TransferStatus[] = ['idle', 'connecting', 'transferring', 'completed', 'error'];
    console.log('✓ Status utility functions work');
    statuses.forEach(status => {
        console.log(`  - ${status}: terminal=${TransferUtils.isTerminalStatus(status)}, active=${TransferUtils.isActiveStatus(status)}`);
    });
    
    // Test type assignments
    const protocol: Protocol = 'tcp';
    const mode: TransferMode = 'transmitter';
    const status: TransferStatus = 'transferring';
    
    console.log('✓ Type assignments work');
    console.log('  - Protocol:', protocol);
    console.log('  - Mode:', mode);
    console.log('  - Status:', status);
    
    console.log('\n🎉 All TypeScript types and utilities tests passed!');
}

// Export for testing
if (typeof window !== 'undefined') {
    // Browser environment - attach to window for manual testing
    (window as any).testTypes = testTypes;
}